arangod/Replication2/ReplicatedLog/Supervision.cpp

```diff
@@ -527,79 +656,161 @@ auto checkReplicatedLog(LogTarget const& target,
-    // doing nothing. Maybe we should have a debug mode where this is still
-    // reported?
-    return EmptyAction("There was nothing to do.");
+    ctx.createAction<ConvergedToTargetAction>(target.version);
+  }
+}
+//
+// This function is called from Agency/Supervision.cpp every k seconds for
+// every replicated log in every database.
+//
+// This means that this function is always going to deal with exactly *one*
+// replicated log.
+//
+// A ReplicatedLog has a Target, a Plan, and a Current subtree in the agency,
+// and these three subtrees are passed into checkReplicatedLog in the form
+// of C++ structs.
+//
+// The return value of this function is an Action, where the type Action is a
+// std::variant of all possible actions that we can perform as a result of
+// checkReplicatedLog.
+//
+// These actions are executes by using std::visit via an Executor struct that
+// contains the necessary context.
+auto checkReplicatedLog(SupervisionContext& ctx, Log const& log,
+                        ParticipantsHealth const& health) -> void {
+  // Check whether the log (that exists in target by virtue of
+  // checkReplicatedLog being called here) is planned. If not, then create it,
+  // provided we have enough participants If there are not enough participants
+  // we can only report back that this log cannot be created.
+  checkLogExists(ctx, log, health);
+
+  checkCurrentExists(ctx, log);
+
+  // Check whether the leader is healthy and all
+  checkLeader(ctx, log, health);
+
+  // Check whether a participant was added in Target that is not in Plan.
+  // If so, add it to Plan.
+  checkParticipantToAdd(ctx, log, health);
+
+  // If a participant is in Plan but not in Target, gracefully
+  // remove it
+  checkParticipantToRemove(ctx, log, health);
+
+  // If the user has updated flags for a participant, which is detected by
+  // comparing Target to Plan, write that change to Plan.
+  // If the configuration differs between Target and Plan,
+  // apply the new configuration.
+  checkParticipantFlagsUpdated(ctx, log, health);
+
+  // TODO: This has not been implemented yet!
+  checkConfigUpdated(ctx, log, health);
+
+  // Check whether we have converged, and if so, report and set version
+  // to target version
+  checkConverged(ctx, log);
+}
+
+// TODO: remove once refactored
+using namespace arangodb::cluster::paths;
+auto executeCheckReplicatedLog(DatabaseID const& dbName,
+                               std::string const& idString, Log log,
+                               ParticipantsHealth const& health,
+                               arangodb::agency::envelope envelope) noexcept
+    -> arangodb::agency::envelope {
+  SupervisionContext ctx;
+
+  checkReplicatedLog(ctx, log, health);
+
+  if (ctx.hasAction()) {
+    envelope = arangodb::replication2::replicated_log::execute(
+        ctx.getAction(), dbName, log.target.id, log.plan, log.current,
+        std::move(envelope));
   }
```

Comment:
Do we want to exclude the `NoActionPossibleAction` here? Same is probably true for the Supervision Actor for the Model Checker?